libcontainer/cgroups/devices/devices_emulator.go

```diff
@@ -79,19 +78,24 @@ func (e *Emulator) IsAllowAll() bool {
 	return e.IsBlacklist() && len(e.rules) == 0
 }
 
-var devicesListRegexp = regexp.MustCompile(`^([abc])\s+(\d+|\*):(\d+|\*)\s+([rwm]+)$`)
-
 func parseLine(line string) (*deviceRule, error) {
-	matches := devicesListRegexp.FindStringSubmatch(line)
-	if matches == nil {
-		return nil, errors.New("line doesn't match devices.list format")
+	// The three fields are: node major:minor perms.
+	fields := strings.Fields(line)
```

Comment:
This will no longer fail to match `[abc]` and `[rwm]`. Does this matter in practice? Probably not, but it is a difference to the regex version.

Reply:
There are checks later in the code for both.

For the first field ("node"), in case it's not one of `a`, `b`, or `c`, the error is returned (line 113).

For the third field ("perms"), `rule.Perms.IsValid()` and `rule.Perms.IsEmpty()` are called to ensure there's nothing but `rwm` (line 142).

So, we're not losing any functionality (it's just the errors are more focused now).

Reply:
Glad that you asked, I looked at the code and found I forgot to remove one `// Should never happen` comment (because now it can happen).